### ISSUE TITLE
Wrong colors using emacs-nox in WT-Ubuntu

### ISSUE THREAD

**Issue by angelog0** [microsoft/terminal#2953]
By default I use MSYS2 MinTTY and by means of `wslbridge` I use also this MinTTY for WSL Ubuntu. With both these apps I use a color scheme which I added to the `profile.json`:

```
{
            "name" : "base16-default-mod-lighten",
            "foreground" : "#D0D0D0",
            "background" : "#151515",
            "black" : "#151515",
            "red" : "#AC4142",
            "green" : "#90A959",
            "yellow" : "#F4BF75",
            "blue" : "#6A9FB5",
            "purple" : "#AA759F",
            "cyan" : "#75B5AA",
            "white" : "#D0D0D0",
            "brightBlack" : "#505050",
            "brightRed" : "#C25E5E",
            "brightGreen" : "#A6BB7B",
            "brightYellow" : "#F8D5A5",
            "brightBlue" : "#8CB5C6",
            "brightPurple" : "#BE95B5",
            "brightCyan" : "#97C7BE",
            "brightWhite" : "#F5F5F5"
        }
```

In the same `profile.json` I added the profiles for MSYS2 bash. I use the same above color scheme for MSYS2 bash and Ubuntu in WT.

In this attachment 
[emacs-nox-sshots.tar.gz](https://github.com/microsoft/terminal/files/3665611/emacs-nox-sshots.tar.gz)
there are a few screenshots of emacs-nox running on various systems. As you can see I have the same good result in MinTTY (emacs-nox-mintty.png), Ubuntu in MinTTY (emacs-nox-mintty-Ubuntu.png) and in MSYS2 bash in WT (emacs-nox-WT-msys2.png). Instead in WT Ubuntu I get what I call _bad_ result (emacs-nox-WT-Ubuntu.png).

**Comment by DHowett-MSFT** [comment]
Can you report what the `TERM` environment variable contains in msys2 and WSL Ubuntu? What do you get when you run WSL ubuntu _outside_ of WT?

**Comment by angelog0** [comment]
@DHowett-MSFT, in WT and outside it (from the Start Menu), Ubuntu has

1.
```
$ echo $TERM
xterm-256color
```

while in MinTTY, Ubuntu has

2.
```
$ echo $TERM
xterm
```

MSYS2 bash in MinTTY has

3.
```
$ echo $TERM
xterm
```

while in WT (using `"commandline": "cmd.exe /c \"set MSYSTEM=MSYS&& C:/msys64/usr/bin/bash.exe --login\""`)

4.
`$ echo $TERM                                                                                                            cygwin `

(here it COPY/PATEs wrong, does not take the `\n`).

In any case, in 2., 3., 4. with the same theme there is the good result. In 1. it is wrong.

BTW, I tested this in WT Ubuntu:

`$ export TERM=xterm `

and the result is similar to 2., 3., 4. : just the comment lines differ, in RED instead of YELLOW.

**Comment by DHowett-MSFT** [comment]
I'm wondering if this is a side effect of #2661.

**Comment by angelog0** [comment]
Just for completeness, not only the colors are wrong but, in short,  Emacs (nox) in WSL in WT DOES NOT work. 

I have a package installed from MELPA, `sr-speedbar`, with which I can select the files to visit in Emacs, just clicking with the mouse, but this does not work. 

Moving the mouse wheel does not scroll the buffer but the content of WT window. 

It is unusable.

The same configuration works just fine in WSL in MinTTY terminal.

**Comment by HoLengZai** [comment]
Any update on that issue... I got the same issue... the colours are horrible, it is really difficult to read it:
I tried others colour schemes but it is the same result. 
![image](https://user-images.githubusercontent.com/20622767/141398789-97a152bb-8fdc-435a-8aa3-47828c6db31a.png)

With a white background, it is a bit better but the font is horrible and painful for the eyes
![image](https://user-images.githubusercontent.com/20622767/141398979-44bbe227-5494-482d-9998-f6575290b670.png)

**Comment by angelog0** [comment]
@HoLengZai , have you tried the color scheme below? It seems a bit better...

```
{
            // From: https://windowsterminalthemes.dev
            "name": "Terminal Basic",
            "black": "#000000",
            "red": "#990000",
            "green": "#00A600",
            "yellow": "#999900",
            "blue": "#0000B2",
            "purple": "#B200B2",
            "cyan": "#00A6B2",
            "white": "#BFBFBF",
            "brightBlack": "#666666",
            "brightRed": "#E50000",
            "brightGreen": "#00D900",
            "brightYellow": "#E5E500",
            "brightBlue": "#0000FF",
            "brightPurple": "#E500E5",
            "brightCyan": "#00E5E5",
            "brightWhite": "#E5E5E5",
            "background": "#FFFFFF",
            "foreground": "#000000",
            "selectionBackground": "#A4C9FF",
            "cursorColor": "#7F7F7F"
        }
```

**Comment by zadjii-msft** [comment]
Okay sorry this one got missed. 

The `xterm` vs `xterm-256color` thing - that makes me think that emacs is using the 16 color palette in mintty, and is using 256-color palette in Windows Terminal. That might also explain why changing the color scheme seems to do nothing in the Terminal - that'd only change the 16 colors, not the rest of the 256 colors. 

I'd suspect that launching emacs with `TERM=xterm emacs-nox` would look more like what you'd want. That's my theory at least. I'm not super familiar with how emacs sets it's own internal color palette unfortunately...

**Comment by HoLengZai** [comment]
> @HoLengZai , have you tried the color scheme below? It seems a bit better...
> 
> ```
> {
>             // From: https://windowsterminalthemes.dev
>             "name": "Terminal Basic",
>             "black": "#000000",
>             "red": "#990000",
>             "green": "#00A600",
>             "yellow": "#999900",
>             "blue": "#0000B2",
>             "purple": "#B200B2",
>             "cyan": "#00A6B2",
>             "white": "#BFBFBF",
>             "brightBlack": "#666666",
>             "brightRed": "#E50000",
>             "brightGreen": "#00D900",
>             "brightYellow": "#E5E500",
>             "brightBlue": "#0000FF",
>             "brightPurple": "#E500E5",
>             "brightCyan": "#00E5E5",
>             "brightWhite": "#E5E5E5",
>             "background": "#FFFFFF",
>             "foreground": "#000000",
>             "selectionBackground": "#A4C9FF",
>             "cursorColor": "#7F7F7F"
>         }
> ```

@angelog0 , Not really, it looks like the same as "One Half Light" default color scheme.
I would like to keep dark/black background, 

As a workaround, I change the default e-macs theme instead of changing the Windows Terminal one:
At least, it is not pain for the eyes and I can read it easily:
![image](https://user-images.githubusercontent.com/20622767/142816417-7264e1f2-2b30-4063-aa89-943a0e0b0875.png)

That's the fastest workaround I guess.
If you want to do the same, I use the built-in "wombat" theme:
`$> emacs ~/.emacs.d/init.el`
and add the following line:
`(load-theme 'wombat t)`

Emacs theme will override all the colours and you will get the same result as above...
If you change the Windows Terminal color scheme, it won't affect Emacs at all, you will only see the window border color changed.

**Comment by angelog0** [comment]
@HoLengZai , thanks for suggestions. I will give it a try...

@msftbot, as you see this thread is not dead...

**Comment by angelog0** [comment]
@HoLengZai , yes with that theme things are better. I tried it with MSYS2 emacs-nox in WT and MinTTY.

In MinTTY it looks a bit different...

**Comment by zadjii-msft** [comment]
Like I said, this sounds like a TERM mismatch. Changing the color scheme of the Terminal isn't going to magically change the colors that emacs emits for it's own scheme, if it's using 256colors for its scheme. You had different `TERM` values for mintty vs Windows Terminal, which would probably explain what you're seeing.

**Comment by HoLengZai** [comment]
@zadjii-msft, @angelog0 

Sorry, I didn't catch your comments @zadjii-msft. (I didn't get that I can use your example as a command on my shell :D, I was probably tired.. or your explanation was a way to advance for my knowledge of terminal customization)
So I read this: [https://github.com/microsoft/terminal/issues/6087](https://github.com/microsoft/terminal/issues/6087)
Then I read this: [https://stackoverflow.com/a/64585/11559874](https://stackoverflow.com/a/64585/11559874)
Then I do some tests: (Alt+M then list-colors-display to see the list of colours supported on emacs)
Then I guess I got your point

**`$>TERM=screen emacs script.sh`**
  <img src="https://user-images.githubusercontent.com/20622767/142911884-f2c1df94-b5ae-41bc-a151-84aa4b8a42a2.png" width="30%" height="30%">

**`$>TERM=xterm emacs script.sh`**
  <img src="https://user-images.githubusercontent.com/20622767/142911891-2c5fef4a-969e-4fc3-a1f8-c6b5a9fd1d4e.png" width="30%" height="30%">

**`$>TERM=xterm-256color emacs script.sh`**
  <img src="https://user-images.githubusercontent.com/20622767/142911899-d9b75327-7012-4f53-b9ea-63698e697bfe.png" width="30%" height="30%">

**`$>TERM=screen-256color emacs script.sh`**
  <img src="https://user-images.githubusercontent.com/20622767/142911919-d1da0bc5-6c2b-4fca-bf80-1c4825e7b5e6.png" width="30%" height="30%">

By default, when I type `env`, I got TERM=xterm-256color
So I guess we can add something like:
`export TERM=xterm`
or
`export TERM=screen-256color`
(which is the 2 best colours scheme, in my opinion, to be able to read properly the source code)

@angelog0 , So I removed "wombat theme" on all my examples above. I just use the default one. Hope it will help anyone like me who just want to have a simple emacs with the proper way to read it through Windows Terminal :D

**Comment by HoLengZai** [comment]
@zadjii-msft 
However, does it mean that Emacs tries to know the term colors used but it doesn't because Windows Terminal does not provide him this info? 

I just try to understand the issue with emacs only even if on environment variable we see xterm-256color....

I found this article for example for iTerm: [https://iterm2.com/faq.html](https://iterm2.com/faq.html)
_Q: How do I get 256 color mode working in vim?
A: The recommended way is to set your terminal type to xterm-256color in Preferences > Profiles > Terminal > Report Terminal Type. You may also need to put set background=light or set background=dark in your .vimrc._

Hence, maybe Windows Terminal can provide this kind of option too? So Emacs will know that we are using xterm-256color

**Comment by zadjii-msft** [comment]
When using WSL, the Terminal (and console) both already report themselves as `xterm-256color`. If you're getting a TERM that's anything else, then it's probably due to running emacs outside of WSL, or something manually changing the `TERM` value. 

I don't understand why emacs would have different color schemes for `xterm-256color` vs `screen-256color`, that doesn't make any dang sense.

**Comment by angelog0** [comment]
@HoLengZai, I did the same tests. With `TERM=screen-256color`, the menu bar has a black background.. It almost does not looks as a menu bar... This shows up also from your images.

I still do not understand why in MinTTY (which has `TERM=xterm`) Emacs colors look better than in WT with the same `TERM` setting...

**Comment by HoLengZai** [comment]
Well, I think the ticket can be closed as it is not specific to Windows Terminal but related to Emacs behaviour to determine the right "terminal (emulator) color"

I finally found the trick to make it work properly without playing with an environment variable or affecting emacs keyboard, etc (as some people got issues by changing/playing with $TERM)
That's the posts that help me to understand/fix the issue:

- [https://stackoverflow.com/questions/7617458/terminal-emacs-colors-only-work-with-term-xterm-256color/18945445#18945445](https://stackoverflow.com/questions/7617458/terminal-emacs-colors-only-work-with-term-xterm-256color/18945445#18945445)
- [https://emacs.stackexchange.com/a/53692](https://emacs.stackexchange.com/a/53692)

So basically if you determine your "frame-background-mode" for emacs, you will be able to fix the issue:

(Using emacs-nox)
```
$>emacs --eval "(setq frame-background-mode 'dark)" script.sh
$>emacs --eval "(setq frame-background-mode 'light)" script.sh
```
Check the one you prefer which match the "normal color"
Mine is dark mode, so I add this entry in my `~/.emacs.d/init.el`
```
(setq frame-background-mode 'dark)
(setq frame-set-background-mode nil) ;;<= I do not find any difference with or without it but i kept it
```

Result: (still using Campbell color scheme)
<img src="https://user-images.githubusercontent.com/20622767/142974230-c81b96bb-1f57-46b3-b195-1ff7798e87f3.png" width="30%" height="30%">

**Comment by zadjii-msft** [comment]
@angelog0 did @HoLengZai's suggestion work for you/?

**Comment by angelog0** [comment]
@zadjii-msft, I do not use dark mode so i cannot say if that works. meanwhile I changed the WSL theme. Now I use `Terminal Basic` with which things with Emacs-nox work better.

Anyway, I think you should discuss these things with Emacs people, the developers...

![wsl-WT-emacs-nox](https://user-images.githubusercontent.com/35931460/184451426-7e1dd2a9-2948-41b9-9f61-e6a4a92e2483.png)

**Comment by angelog0** [comment]
BTW, MSYS2 has a TTY package of [Emacs](https://packages.msys2.org/base/emacs) and to use it in WT I had to chance Emacs theme:

```
;; See https://github.com/microsoft/terminal/issues/2953#issuecomment-975180916
(load-theme 'wombat t)
```

With that theme, things work better. **BETTER** does not mean **RIGHT**! For this reason, in previous post, I suggested to discus the issue with Emacs people.
